lib/pacemaker/pcmk_output.c

```diff
@@ -2344,6 +2344,237 @@ result_code_xml(pcmk__output_t *out, va_list args)
+
+    pcmk__output_xml_create_parent(out, PCMK_XE_TICKET,
+                                   PCMK_XA_ID, ticket_id, NULL);
+    pcmk__output_xml_create_parent(out, PCMK_XE_CONSTRAINTS, NULL);
+    pcmk__output_xml_add_node_copy(out, node);
+
+    /* Pop two parents so now we are back under the <tickets> element */
+    pcmk__output_xml_pop_parent(out);
+    pcmk__output_xml_pop_parent(out);
+
+    return pcmk_rc_ok;
+}
+
+static int
+add_resource_element(xmlNode *node, void *userdata)
+{
+    pcmk__output_t *out = (pcmk__output_t *) userdata;
+    const char *rsc = crm_element_value(node, PCMK_XA_RSC);
+
+    pcmk__output_create_xml_node(out, PCMK_XE_RESOURCE,
+                                 PCMK_XA_ID, rsc, NULL);
+    return pcmk_rc_ok;
+}
+
+PCMK__OUTPUT_ARGS("ticket-constraints", "xmlNode *")
+static int
+ticket_constraints_xml(pcmk__output_t *out, va_list args)
+{
+    xmlNode *constraint_xml = va_arg(args, xmlNode *);
+
+    /* Create:
+     * <tickets>
+     *   <ticket id="">
+     *     <constraints>
+     *       <rsc_ticket />
+     *     </constraints>
+     *   </ticket>
+     *   ...
+     * </tickets>
+     */
+    pcmk__output_xml_create_parent(out, PCMK_XE_TICKETS, NULL);
+
+    if (pcmk__xe_is(constraint_xml, PCMK__XE_XPATH_QUERY)) {
+        /* Iterate through the list of children once to create all the
+         * ticket/constraint elements.
+         */
+        pcmk__xe_foreach_child(constraint_xml, NULL, add_ticket_element_with_constraints, out);
+
+        /* Put us back at the same level as where <tickets> was created. */
+        pcmk__output_xml_pop_parent(out);
+
+        /* Constraints can reference a resource ID that is defined in the XML
+         * schema as an IDREF.  This requires some other element to be present
+         * with an id= attribute that matches.
+         *
+         * Iterate through the list of children a second time to create the
+         * following:
+         *
+         * <resources>
+         *   <resource id="" />
+         *   ...
+         * </resources>
+         */
+        pcmk__output_xml_create_parent(out, PCMK_XE_RESOURCES, NULL);
+        pcmk__xe_foreach_child(constraint_xml, NULL, add_resource_element, out);
+        pcmk__output_xml_pop_parent(out);
+
+    } else {
+        /* Creating the output for a single constraint is much easier.  All the
+         * comments in the above block apply here.
+         */
+        add_ticket_element_with_constraints(constraint_xml, out);
+        pcmk__output_xml_pop_parent(out);
+
+        pcmk__output_xml_create_parent(out, PCMK_XE_RESOURCES, NULL);
+        add_resource_element(constraint_xml, out);
+        pcmk__output_xml_pop_parent(out);
+    }
+
+    return pcmk_rc_ok;
+}
+
+PCMK__OUTPUT_ARGS("ticket-state", "xmlNode *")
+static int
+ticket_state_default(pcmk__output_t *out, va_list args)
+{
+    xmlNode *state_xml = va_arg(args, xmlNode *);
+
+    GString *buf = g_string_sized_new(1024);
+
+    out->info(out, "State XML:\n");
```

Comment:
We generally don't want newlines with `info()` since it could be unpredictable with different formats. Maybe just add a `spacer()` call. But that can be done with the next PR